### PR TITLE
Bug fix for the datetime in Smap converter

### LINKED ISSUE #1023: SMAP converter produces the wrong datetime
The converted datetime looks off. Discovered while adding smap to the cycling.

### PR DESCRIPTION
**Pull request by apchoiCMD**
#### For Smap_SSS:
- Replace the method to get a correct datetime from epoch time by using boost/date_time/Gregorian
- JulianDate - 2440588 will make half day off- 
- The above method will avoid dealing with decimal precision.
- Resolves #1023

### CODE REVIEW

File: utils/obsproc/Smap2Ioda.h [hunk 1]
```diff
@@ -83,21 +85,47 @@ namespace gdasapp {
...
+      while (day > 31) {
+          if (month == 4 || month == 6 || month == 9 || month == 11) {
+              day -= 30;
+          } else if (month == 2) {
+              day -= 28;  // Assuming not a leap year
+          } else {
+              day -= 31;
+          }
+          month++;
+      }
```
Comment (guillaumevernieres) [inline utils/obsproc/Smap2Ioda.h]: That looks wrong (what if it is a leap year) and definitely out of place.

File: utils/obsproc/Smap2Ioda.h [hunk 2]
```diff
@@ -83,21 +85,47 @@ namespace gdasapp {
...
-      int daysSinceEpoch = julianDate - 2440588;
+      // Set epoch time for SMAP_SSS
+      util::DateTime epochDtime("1970-01-01T00:00:00Z");
 
-      // Calculate seconds
-      int secondsSinceEpoch = (daysSinceEpoch + startDay) * 86400;
+      // Convert Obs DateTime objects to epoch time offsets in seconds
+      // 0000-00-00T00:00:00Z will be converterd to negative seconds
+      int64_t timeOffsets
+           = ioda::convertDtimeToTimeOffsets(epochDtime, {dateTime})[0];
```
Comment (guillaumevernieres) [inline utils/obsproc/Smap2Ioda.h:103]: we shouldn't have to do this.
Try to make use of boost:
```c++
#include <boost/date_time/gregorian/gregorian.hpp>
```
has a bunch of nifty datetime utilities that probably would have saved us some headache.

File: utils/obsproc/Smos2Ioda.h [hunk 3]
```diff
@@ -80,7 +80,7 @@ namespace gdasapp {
         iodaVars.obsVal_(i) = sss[i];
         iodaVars.obsError_(i) = sss_error[i];
         iodaVars.preQc_(i) = sss_qc[i];
-        iodaVars.datetime_(i) =  static_cast<int64_t>(datetime_[i]*86400.0f) + mjd2000;
+        iodaVars.datetime_(i) =  static_cast<int64_t>((datetime_[i]*86400.0f) + mjd2000);
```
Comment (guillaumevernieres) [inline utils/obsproc/Smos2Ioda.h]: is there a point to the addition of the parenthesis?

**Review by guillaumevernieres — CHANGES_REQUESTED** [review]
We shouldn't have to worry about leap years ... Somebody did before us. I think this can be re-written using ```boost```

**Review by AndrewEichmann-NOAA — COMMENTED** [review]
Nothing to add - yeah there should be libraries

File: utils/obsproc/Smos2Ioda.h (hunk 3 above)
Comment (apchoiCMD) [reply utils/obsproc/Smos2Ioda.h]: No, it is correct and I reverse smos.

File: utils/obsproc/Smap2Ioda.h [hunk 4]
```diff
@@ -82,22 +84,24 @@ namespace gdasapp {
...
-        year = 0;
+        // 1400 is given by boost library
+        // final seconds goes negative then trimmed
+        year = 1400;
       }
 
-      // Convert a date to Julian date
-      uint64_t julianDate = util::datefunctions::dateToJulian(year, month, day);
+      // Define a date
+      boost::gregorian::date d(year, 1, 1);
```
Comment (guillaumevernieres) [inline utils/obsproc/Smap2Ioda.h]: could we try to avoid single letter variables ...

File: utils/obsproc/Smap2Ioda.h [hunk 5]
```diff
@@ -82,22 +84,24 @@ namespace gdasapp {
...
+
+      // Add a startDay to the defined date
```
Comment (guillaumevernieres) [inline utils/obsproc/Smap2Ioda.h]: that does not describe what's done below

File: utils/obsproc/Smap2Ioda.h [hunk 6]
```diff
@@ -82,22 +84,24 @@ namespace gdasapp {
...
+      d += boost::gregorian::date_duration(startDay - 1);
```
Comment (guillaumevernieres) [inline utils/obsproc/Smap2Ioda.h]: why do we need to do this?

**Review by guillaumevernieres — COMMENTED** [review]
Some minor clean up and that should be good to go. Thanks @apchoiCMD .

File: utils/obsproc/Smap2Ioda.h [hunk 7]
```diff
@@ -82,22 +84,24 @@ namespace gdasapp {
...
       if (year == -9999) {
-        year = 0;
+        // 1400 is given by boost library
+        // final seconds goes negative then trimmed
+        year = 1400;
       }
 
-      // Convert a date to Julian date
-      uint64_t julianDate = util::datefunctions::dateToJulian(year, month, day);
+      // Define a date
```
Comment (guillaumevernieres) [inline utils/obsproc/Smap2Ioda.h]: sorry for being picky with comments, but what do you mean by "define a date"? A random date?

**Review by AndrewEichmann-NOAA — APPROVED** [review]
This looks much nicer

**Review by guillaumevernieres — APPROVED** [review]
Thanks @apchoiCMD . Did you create a github issue for the change in datetime format prior to 2021 (or whatever the transition date is)?

**Pull request merged**